File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/service/CategoryService.java
package com.example.ecommerce.service;

import com.example.ecommerce.dto.CategoryDto;
import com.example.ecommerce.entity.Category;
import com.example.ecommerce.exception.ResourceNotFoundException;
import com.example.ecommerce.repository.CategoryRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class CategoryService {

    private final CategoryRepository categoryRepository;

    public CategoryService(CategoryRepository categoryRepository) {
        this.categoryRepository = categoryRepository;
    }

    @Transactional(readOnly = true)
    public List<CategoryDto> findAll() {
        return categoryRepository.findAll().stream().map(this::toDto).toList();
    }

    @Transactional(readOnly = true)
    public CategoryDto findById(Long id) {
        return toDto(getCategory(id));
    }

    @Transactional
    public CategoryDto create(CategoryDto dto) {
        Category category = new Category();
        category.setName(dto.name());
        category.setDescription(dto.description());
        return toDto(categoryRepository.save(category));
    }

    public Category getCategory(Long id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found: " + id));
    }

    private CategoryDto toDto(Category category) {
        return new CategoryDto(category.getId(), category.getName(), category.getDescription());
    }
}

File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Services/CategoryService.cs
using Ecommerce.Dto;
using Ecommerce.Models;
using Ecommerce.Exceptions;
using Ecommerce.Repositories;

namespace Ecommerce.Services;

public class CategoryService
{
    private readonly CategoryRepository _categoryRepository;

    public CategoryService(CategoryRepository categoryRepository)
    {
        _categoryRepository = categoryRepository;
    }

    public async Task<List<CategoryDto>> FindAllAsync()
    {
        var categories = await _categoryRepository.FindAllAsync();
        return categories.Select(ToDto).ToList();
    }

    public async Task<CategoryDto> FindByIdAsync(long id)
    {
        return ToDto(await GetCategoryAsync(id));
    }

    public async Task<CategoryDto> CreateAsync(CategoryDto dto)
    {
        var category = new Category
        {
            Name = dto.Name,
            Description = dto.Description
        };
        category = await _categoryRepository.SaveAsync(category);
        return ToDto(category);
    }

    public async Task<Category> GetCategoryAsync(long id)
    {
        return await _categoryRepository.FindByIdAsync(id)
            ?? throw new ResourceNotFoundException($"Category not found: {id}");
    }

    private static CategoryDto ToDto(Category category) =>
        new(category.Id, category.Name, category.Description);
}
